### PR TITLE
Add aten::nll_loss_backward op

### PR DESCRIPTION
**Pull request by pashu123**
The lowering of aten::nll_loss_backward op has been added
from torch to linalg dialect. The changes has been made as
a part of -torch-convert-to-linalg pass.

Signed-off-by: Prashant Kumar <prashant@nod-labs.com>

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 1]
```diff
@@ -966,6 +971,27 @@ ChangeResult TypeAnalyzer::visitAtenNllLossForwardOp(
...
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+
+  knowledge.dtype = self.dtype;
+  int64_t reduction;
+  unsigned resultRank = self.sizes.size();
+
+  if (self.hasSizes &&
+      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+    // reduction == 1 means reduce 1st dim.
+    resultRank = reduction == 1 ? resultRank - 1 : resultRank;
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:988]: I don't think this is true. From the definition

https://github.com/pytorch/pytorch/blob/617a3bd944320d2d40bae78aefbfcdebf5255a59/aten/src/ATen/native/LossNLL.cpp#L110

`NllLossBackward` returns a tensor of the same size as the `self` argument.

Also, the visitor for `NllLossForward`, where this calculation is also used, does not seem right to me. The case of `reduction == 2` is not handled, and it is not clear to me that subtracting one from the result rank results in a scalar, as specified by the documentation

https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Thanks for this! A few comments

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:988]: Thanks for the comments, I will take a look at this. I guess I have made some mistakes.

Comment (pashu123) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:1221]: So, the shape of output is the same as the shape of the self in the case of NLL loss backward and I have modified the case of NLL loss forward taking into account both reductions.

Comment (pashu123) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:1221]: So, the shape of output is the same as the shape of the self in the case of NLL loss backward and I have modified the case of NLL loss forward taking into account both reductions.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 2]
```diff
@@ -1263,6 +1263,107 @@ class ConvertAtenNllLossForwardOp
...
+                  // grad_input[i][j] = (j == target[i]) ? grad_output[i] : 0
+                  Value cmpEq = rewriter.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, indJ, indTarget);
+
+                  // The target index shouldn't be equal to `ignoreIndex`.
+                  Value cmpNEq = rewriter.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::ne, ignoreIndexVal, indTarget);
+                  Value finalPredicate =
+                      rewriter.create<arith::AndIOp>(loc, cmpEq, cmpNEq);
+                  Value negate =
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1419]: Updated.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp]: scalar -> 0d tensor?

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 3]
```diff
@@ -1323,6 +1328,108 @@ class ConvertAtenNllLossForwardOp
...
+  LogicalResult
+  matchAndRewrite(AtenNllLossBackwardOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+    Location loc = op->getLoc();
+    Value input = adaptor.self();
+    Value target = adaptor.target();
+    Value weight = adaptor.weight();
+    Value grad_output = adaptor.grad_output();
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1350]: nit: gradOuput

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 4]
```diff
@@ -1323,6 +1328,108 @@ class ConvertAtenNllLossForwardOp
...
+                      loc, rewriter.getIndexType(), args[0]);
+                  Value indJ = rewriter.create<linalg::IndexOp>(loc, 1);
+
+                  // The final result is given by:
+                  // grad_input[i][j] = (j == target[i]) ? -grad_output[i] : 0
+                  Value cmpEq = rewriter.create<arith::CmpIOp>(
+                      loc, arith::CmpIPredicate::eq, indJ, indTarget);
+
+                  // The target index shouldn't be equal to `ignoreIndex`.
+                  Value cmpNEq = rewriter.create<arith::CmpIOp>(
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1415]: nit: cmpNe

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 5]
```diff
@@ -328,6 +328,11 @@ static Value buildUnitNormalCdf(OpBuilder &b, Location &loc, Value x) {
   return buildNormalCdf(b, loc, x, zero, one);
 }
 
+// These constants control the reduction behavior of the loss functions.
+// None, Mean and Sum corresponds to "do not reduce", "Mean of losses", and "sum
+// of losses" respectively.
+enum Reduction { None, Mean, Sum, END };
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:334]: we can move this definition to `include/torch-mlir/Dialect/Torch/Utils/Utils.h` since it's shared by multiple files. Also, please include the `aten/src/ATen/core/Reduction.h` in the comments for future reference.

**Review by cathyzhyi — APPROVED** [review]

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 5 above)
Comment (ramiro050) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:334]: Would the file https://github.com/llvm/torch-mlir/blob/main/include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h be a better location for this `enum`, since it is code from torch upstream?

Comment (cathyzhyi) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:334]: yea, you are right. that file should be a better place.

File: include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h [hunk 6]
```diff
@@ -70,6 +70,13 @@ struct ResultTypeState {
 ScalarType result_type(const ResultTypeState &in_state);
 ScalarType promote_skip_undefined(ScalarType a, ScalarType b);
 
+// These constants control the reduction behavior of the loss functions.
```
Comment (cathyzhyi) [inline include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h:73]: nit: can you change the comment to similar format as the other comments?

Comment (pashu123) [reply include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h:74]: Done, thanks.
